homeassistant/components/icloud.py

```diff
@@ -0,0 +1,946 @@
+"""
+Platform that supports scanning iCloud.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/icloud/
+"""
+import logging
+from datetime import datetime, timedelta
+from math import floor
+import random
+import re
+
+from pytz import timezone
+import pytz
+
+from homeassistant.const import CONF_USERNAME, CONF_PASSWORD
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.entity import generate_entity_id
+from homeassistant.components.device_tracker import see
+from homeassistant.helpers.event import track_state_change
+from homeassistant.helpers.event import track_utc_time_change
+import homeassistant.util.dt as dt_util
+from homeassistant.util.location import distance
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ['pyicloud==0.9.1']
+
+DEPENDENCIES = ['zone', 'device_tracker']
+
+CONF_EVENTS = 'events'
+DEFAULT_EVENTS = False
+
+CONF_COOKIEDIRECTORY = 'cookiedirectory'
+
+# entity attributes
+ATTR_ACCOUNTNAME = 'accountname'
+ATTR_INTERVAL = 'interval'
+ATTR_DEVICENAME = 'devicename'
+ATTR_BATTERY = 'battery'
+ATTR_DISTANCE = 'distance'
+ATTR_STARTTIME = 'start_time'
+ATTR_ENDTIME = 'end_time'
+ATTR_DURATION = 'duration'
+ATTR_REMAINING = 'remaining_time'
+ATTR_DEVICESTATUS = 'device_status'
+ATTR_LOWPOWERMODE = 'low_power_mode'
+ATTR_BATTERYSTATUS = 'battery_status'
+ATTR_LOCATION = 'location'
+ATTR_FRIENDLY_NAME = 'friendly_name'
+ATTR_GOOGLE_MAPS_TRAVEL_TIME = 'google_maps_travel_time'
+ATTR_GOOGLE_MAPS_TRAVEL_TIME_DURATION = 'gmtt_duration'
+ATTR_GOOGLE_MAPS_TRAVEL_TIME_ORIGIN = 'gmtt_origin'
+
+TYPE_CURRENT = 'currentevent'
+TYPE_NEXT = 'nextevent'
+
+ICLOUDTRACKERS = {}
+
+DOMAIN = 'icloud'
+DOMAIN2 = 'idevice'
+DOMAIN3 = 'ievent'
+
+ENTITY_ID_FORMAT_ICLOUD = DOMAIN + '.{}'
+ENTITY_ID_FORMAT_DEVICE = DOMAIN2 + '.{}'
+ENTITY_ID_FORMAT_EVENT = DOMAIN3 + '.{}'
+
+DEVICESTATUSSET = ['features', 'maxMsgChar', 'darkWake', 'fmlyShare',
```

Comment:
How would you suggest I make the PR upstream? I could only add each name to the pyicloud repo, because the set I defined here has all the available attributes for the iDevices, but other programs might not want all this data, so this can't be a fixed set in there :s